man/man1/webgen.1
.TH "WEBGEN" 1 "September 2013"
.SH NAME
webgen \- a fast, powerful and extensible static website generator
.SH SYNOPSIS
.B webgen
[\fIglobal options\fR]
\fIcommand\fR
[\fIcommand arguments\fR] [command [command arguments] ...] [args]
.SH DESCRIPTION
webgen is used to generate static websites from content files written in a markup language and
template files. You create one template file in which you define the layout of your pages and where
the content should go. After that you create content files called page files in which you only
define the content. webgen merges the page's content into the template and generates the final HTML
output. During the last step so called webgen tags are substituted so that, for example, menus can
automatically be generated.
.SH GLOBAL OPTIONS
.TP
.B \-d, \-\-directory DIR
Set the website directectory. if none specified, the current directory is used.
.TP
.B \-s, \-\-search-dirs
Search the parent directories for a directory with a webgen configuration file and use this as
website directory.
.TP
.B \-c, \-\-color
Use ANSI colors for the output.
.TP
.B \-n, \-\-dry-run
Simulate the website generation. Does everything as it would be done without the option but nothing
gets written to the destination directory.
.TP
.B \-o, \-\-option CONFIG_OPTION
Specify a simple configuration option. The argument must be of the form KEY=VALUE where the VALUE is
a valid YAML type.
.TP
.B \-v, \-\-verbose
Show more output by default.
.TP
.B \-q, \-\-quiet
Show less output by default. If this option is specified, then no output is shown during rendering
of a website.
.TP
.B \-\-debug
Enable debugging (shows much more output).
.TP
.B \-h, \-\-help
Show generic help.
.TP
.B \-V, \-\-version
Show the version of the program and exit.
.SH COMMANDS
Commands are used to execute different actions (e.g. creating a website directory, generating the
website, showing help, ...). The command names can be shortened as long as the used name is
unambiguous. For example, the command "show deps" can also be executed by specifying "s d".
.TP
\fBcreate bundle\fR
Creates a new extension bundle either as a standalone extension or inside the website extension directory.
.TP
\fBcreate website\fR
Creates the basic directories and files for webgen. This includes the source and destination
directories as well as the extension directory. Takes the \fIdirectory\fR in which to create the
webgen structure as an argument.
.TP
\fBgenerate\fR
Generates the website. This command will be executed if no command is explicitly specified.
.TP
\fBhelp\fR
Provides help for individual commands.
.TP
\fBinstall\fR
Installs an extension bundle. The name of the extension bundle is needed as argument.
.TP
\fBshow bundles\fR
Shows all available bundles and additional information them.
.TP
\fBshow config\fR
Shows all available configuration options. Command options can be used to additionally show default
values and/or current values.
.TP
\fBshow deps\fR
Shows dependencies of nodes. After webgen generates the website (see generate) information about the
generated website is stored in cache files and this command reads this information to show the
dependencies of nodes.
.TP
\fBshow extensions\fR
Shows all available extensions.
.TP
\fBshow tree\fR
Shows the node tree. Command options can be used to customize the output.
.TP
\fBversion\fR Shows the version of the program and exit.
.SH ENVIRONMENT VARIABLES
webgen's behaviour is affected by the following environment variables:
.TP
\fBWEBGEN_WEBSITE\fR
If no website is specified via the \-d option and this variable is set, webgen assumes that its
content is the webgen website directory that should be used.
.SH SEE ALSO
The webgen website, http://webgen.gettalong.org/ for detailed information.
.SH AUTHOR
webgen was written by Thomas Leitner <t_leitner@gmx.at>
.PP
This manual page was written by Paul van Tilburg <paulvt@debian.org> for the Debian project (but may
be used by others).  This manual page was update to reflect webgen 0.4 change by Arnaud Cornet
<arnaud.cornet@gmail.com>.  This manual page was updated to reflect webgen 0.5.x changes by Thomas
Leitner <t_leitner@gmx.at>.  This manual page was updated to reflect webgen 1.x changes by Thomas
Leitner <t_leitner@gmx.at>.



lib/webgen.rb
#
#--
#
# $Id$
#
# webgen: template based static website generator
# Copyright (C) 2004 Thomas Leitner
#
# This program is free software; you can redistribute it and/or modify it under the terms of the GNU
# General Public License as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with this program; if not,
# write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#
#++
#

require 'rbconfig'
require 'fileutils'
require 'cmdparse'
require 'webgen/website'
require 'webgen/plugin'

module Webgen

  class Color

    @@colors = {:bold => [0, 1], :green => [0, 32], :lblue => [1, 34], :lred => [1, 31], :reset => [0, 0]}

    def self.colorify
      @@colors.each do |color, values|
        module_eval <<-EOF
        def Color.#{color.to_s}
        "\e[#{values[0]};#{values[1]}m"
        end
        EOF
      end
    end

    def self.method_missing( id )
      ''
    end

  end

  class CreateCommand < CmdParse::Command

    def initialize
      super( 'create', false )
      self.short_desc = "Creates the basic directories and files for webgen"
      self.description = "You can optionally specify a template and/or a style which should be used. " \
      "This allows you to create a good starting template for your website."
      self.options = CmdParse::OptionParserWrapper.new do |opts|
        opts.separator "Options:"
        opts.on( '-t', '--template TEMPLATE', Webgen::WebSiteTemplate.entries.keys, 'Specify the template which should be used' ) {|@template|}
        opts.on( '-s', '--style STYLE', 'Specify the style which should be used' ) {|@style|}
        opts.separator ""
        opts.separator "Available templates and styles:"
        opts.separator ""
        opts.separator opts.summary_indent + "#{Color.bold}Templates#{Color.reset}"
        Webgen::WebSiteTemplate.entries.sort.each do |name, entry|
          opts.separator opts.summary_indent + "  #{Color.green}Name:#{Color.reset}        #{Color.lred}#{name}#{Color.reset}"
          opts.separator opts.summary_indent + "  #{Color.green}Description:#{Color.reset} #{entry.infos['description']}"
          opts.separator ""
        end
        opts.separator opts.summary_indent + "#{Color.bold}Styles#{Color.reset}"
        Webgen::WebSiteStyle.entries.sort.each do |name, entry|
          opts.separator opts.summary_indent + "  #{Color.green}Name:#{Color.reset}        #{Color.lred}#{name}#{Color.reset}"
          opts.separator opts.summary_indent + "  #{Color.green}Description:#{Color.reset} #{entry.infos['description']}"
          opts.separator ""
        end
      end
      @template = 'default'
      @style = 'default'
    end

    def usage
      "Usage: #{commandparser.program_name} [global options] create [options] DIR"
    end

    def execute( args )
      if args.length == 0
        raise OptionParser::MissingArgument.new( 'DIR' )
      else
        Webgen.create_website( args[0], @template, @style )
      end
    end

  end

  class CleanCommand < CmdParse::Command

    module ::FileUtils

      def fu_output_message( msg )
        logger.info { msg }
      end

      def ask_before_delete( ask, func, list, options = {} )
        newlist = [*list].collect {|e| "'" + e + "'"}.join(', ')
        if ask
          print "Really delete #{newlist}? "
          return unless /y|Y/ =~ gets.strip
        end
        self.send( func, list, options.merge( :verbose => true ) )
      end

      module_function :ask_before_delete

    end

    class CleanWalker < Webgen::Plugin

      summary "Deletes the output file for each node."
      depends_on 'TreeWalker'

      attr_writer :ask

      def handle_node( node, level )
        file = node.recursive_value( 'dest' )
        return if !File.exists?( file ) || node['int:virtualNode']
        if File.directory?( file )
          begin
            FileUtils.ask_before_delete( @ask, :rmdir, file )
          rescue Errno::ENOTEMPTY => e
            logger.info "Cannot delete directory #{file}, as it is not empty!"
          end
        else
          FileUtils.ask_before_delete( @ask, :rm, file, :force => true )
        end
      end

    end

    def initialize;
      super( 'clean', false )
      self.short_desc = "Removes the generated or all files from the output directory"
      self.options = CmdParse::OptionParserWrapper.new do |opts|
        opts.separator( "Options" )
        opts.on( '--all', '-a', "Removes ALL files from the output directory" ) { @all = true }
        opts.on( '--interactive', '-i', "Ask for each file" ) { @ask = true }
      end
      @all = false
      @ask = false
    end

    def execute( args )
      if @all
        logger.info( "Deleting all files from output directory..." )
        outDir = Plugin['Configuration']['outDirectory']
        FileUtils.ask_before_delete( @ask, :rm_rf, outDir ) if File.exists?( outDir )
      else
        tree = Plugin['FileHandler'].build_tree
        logger.info( "Deleting generated files from output directory..." )
        Plugin['CleanWalker'].ask = @ask
        Plugin['TreeWalker'].execute( tree, Plugin['CleanWalker'], :backward )
      end
      logger.info( "Webgen finished 'clean' command" )
    end

  end

  class WebgenCommandParser < CmdParse::CommandParser

    def initialize
      super( true )
      self.program_name = "webgen"
      self.program_version = Webgen::VERSION
      self.options = CmdParse::OptionParserWrapper.new do |opts|
        opts.separator "Global options:"
        opts.on( "--verbosity LEVEL", "-V", Integer, "The verbosity level" ) { |verbosity| Plugin.set_param( 'Logging', 'verbosityLevel', verbosity ) }
        opts.on( "--[no-]logfile", "-L", "Log to file" ) { |logfile| Plugin.set_param( 'Logging', 'logToFile', logfile ) }
      end

      # Run command
      run = CmdParse::Command.new( 'run', false )
      run.short_desc = "Runs webgen"
      run.set_execution_block {|args| Webgen.run_webgen }
      self.add_command( run, true )

      # Show command
      show = CmdParse::Command.new( 'show', true )
      show.short_desc = "Show various information"
      self.add_command( show )

      # Show plugins command
      showPlugins = CmdParse::Command.new( 'plugins', false )
      showPlugins.short_desc = "Show the available plugins"
      showPlugins.set_execution_block do |args|
        print "List of loaded plugins:\n"

        headers = Hash.new {|h,k| h[k] = (k.nil? ? "Other Plugins" : k.gsub(/([A-Z][a-z])/, ' \1').strip) }

        ljustlength = 30 + Color.green.length + Color.reset.length
        header = ''
        Plugin.config.sort { |a, b| a[0].name <=> b[0].name }.each do |klass, data|
          newHeader = headers[klass.name[/^.*?(?=::)/]]
          unless newHeader == header
            print "\n  #{Color.bold}#{newHeader}#{Color.reset}:\n";
            header = newHeader
          end
          print "    - #{Color.green}#{data.plugin}#{Color.reset}:".ljust(ljustlength) +"#{data.summary}\n"
        end
      end
      show.add_command( showPlugins )

      # Show params command
      showConfig = CmdParse::Command.new( 'config', false )
      showConfig.short_desc = "Show the parameters of all plugins or of the specified one"
      showConfig.set_execution_block do |args|
        puts "List of plugin parameters:"
        puts
        ljustlength = 25 + Color.green.length + Color.reset.length
        Plugin.config.sort { |a, b| a[0].name[/\w*$/] <=> b[0].name[/\w*$/] }.each do |klass, data|
          next if args.length > 0 && args[0] != data.plugin
          puts "  #{Color.bold}#{data.plugin}#{Color.reset}:"
          puts "    #{Color.green}Summary:#{Color.reset}".ljust(ljustlength) + data.summary if data.summary
          puts "    #{Color.green}Description:#{Color.reset}".ljust(ljustlength) + data.description if data.description
          puts "    #{Color.green}Tag name:#{Color.reset}".ljust(ljustlength) + (data.tag == :default ? "Default tag" : data.tag) if data.tag
          puts "    #{Color.green}Handled paths:#{Color.reset}".ljust(ljustlength) + data.path.inspect if data.path

          data.table.keys.find_all {|k| /^registered_/ =~ k.to_s }.each do |k|
            name = k.to_s.sub( /^registered_/, '' ).tr('_', ' ').capitalize + " name"
            puts "    #{Color.green}#{name}:#{Color.reset}".ljust(ljustlength) + data.send( k )
          end

          if data.params
            puts "\n    #{Color.green}Parameters:#{Color.reset}"
            data.params.sort.each do |key, item|
              print "      #{Color.green}Parameter:#{Color.reset}".ljust(ljustlength)
              puts Color.lred + item.name + Color.reset + " = " + Color.lblue +  item.value.inspect + Color.reset + " (" + item.default.inspect + ")"
              puts "      #{Color.green}Description:#{Color.reset}".ljust(ljustlength) + item.description
              puts
            end
          end
          puts
        end
      end
      show.add_command( showConfig )

      self.add_command( CreateCommand.new )
      self.add_command( CleanCommand.new )
      self.add_command( CmdParse::HelpCommand.new )
      self.add_command( CmdParse::VersionCommand.new )
    end

  end

  # Main program for the webgen CLI command.
  def self.cli_main
    Color.colorify if $stdout.isatty && !Config::CONFIG['arch'].include?( 'mswin32' )
    Plugin['Configuration'].cmdparser = WebgenCommandParser.new
    Plugin['Configuration'].cmdparser.parse do |level, cmdName|
      Plugin['Configuration'].init_all if level == 0
    end
  end

end
